File: scripts/fetch_browser_rendered_sources.mjs
import { createRequire } from "node:module";
import { createHash } from "node:crypto";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { mkdir, readFile, writeFile } from "node:fs/promises";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));
const DEFAULT_QUEUE = join(ROOT, "data", "mvp", "browser_render_queue.csv");
const DEFAULT_CSV_OUT = join(ROOT, "data", "extraction", "browser_rendered_sources.csv");
const DEFAULT_JSON_OUT = join(ROOT, "data", "extraction", "browser_rendered_sources.json");
const DEFAULT_TEXT_DIR = join(ROOT, "data", "extraction", "rendered_source_texts");
const MIN_TEXT_LENGTH = 500;

function parseArgs(argv) {
  const args = {
    queue: DEFAULT_QUEUE,
    csvOut: DEFAULT_CSV_OUT,
    jsonOut: DEFAULT_JSON_OUT,
    limit: undefined,
    timeoutMs: 30000,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    const next = argv[index + 1];
    if (arg === "--queue") args.queue = next;
    if (arg === "--csv-out") args.csvOut = next;
    if (arg === "--json-out") args.jsonOut = next;
    if (arg === "--limit") args.limit = Number.parseInt(next, 10);
    if (arg === "--timeout-ms") args.timeoutMs = Number.parseInt(next, 10);
    if (arg.startsWith("--")) index += 1;
  }
  return args;
}

function parseCsv(text) {
  const rows = [];
  let field = "";
  let row = [];
  let quoted = false;
  const pushField = () => {
    row.push(field);
    field = "";
  };
  const pushRow = () => {
    if (row.length === 1 && row[0] === "") return;
    rows.push(row);
    row = [];
  };

  for (let index = 0; index < text.length; index += 1) {
    const char = text[index];
    const next = text[index + 1];
    if (quoted) {
      if (char === '"' && next === '"') {
        field += '"';
        index += 1;
      } else if (char === '"') {
        quoted = false;
      } else {
        field += char;
      }
    } else if (char === '"') {
      quoted = true;
    } else if (char === ",") {
      pushField();
    } else if (char === "\n") {
      pushField();
      pushRow();
    } else if (char !== "\r") {
      field += char;
    }
  }
  pushField();
  pushRow();

  const [headers, ...dataRows] = rows;
  if (!headers) return [];
  return dataRows.map((values) =>
    Object.fromEntries(headers.map((header, index) => [header, values[index] ?? ""])),
  );
}

function csvEscape(value) {
  const text = String(value ?? "");
  if (/[",\n\r]/.test(text)) return `"${text.replaceAll('"', '""')}"`;
  return text;
}

async function writeOutputs(rows, csvOut, jsonOut) {
  const fieldnames = [
    "rank",
    "mvp_scope",
    "brand",
    "model",
    "registrations_ytd",
    "source_url",
    "final_url",
    "http_status",
    "source_domain",
    "render_status",
    "render_error",
    "source_hash",
    "source_text_path",
    "source_text_length",
    "source_text_preview",
    "fetched_at",
    "extraction_status",
    "notes",
  ];
  const csv = [
    fieldnames.join(","),
    ...rows.map((row) => fieldnames.map((field) => csvEscape(row[field])).join(",")),
  ].join("\n") + "\n";
  await mkdir(dirname(csvOut), { recursive: true });
  await mkdir(dirname(jsonOut), { recursive: true });
  await writeFile(csvOut, csv, "utf8");
  await writeFile(jsonOut, `${JSON.stringify(rows, null, 2)}\n`, "utf8");
}

function mergeRenderedRows(existingRows, newRows) {
  const rows = new Map();
  for (const row of existingRows) {
    if (row.source_url) rows.set(row.source_url, row);
  }
  for (const row of newRows) {
    if (row.source_url) rows.set(row.source_url, row);
  }
  return [...rows.values()].sort((a, b) => Number(a.rank || 999) - Number(b.rank || 999));
}

function loadPlaywright() {
  const attempts = [
    createRequire(import.meta.url),
    createRequire(join(homedir(), ".cache", "codex-runtimes", "codex-primary-runtime", "dependencies", "node", "node_modules", "playwright", "package.json")),
  ];
  const errors = [];
  for (const requireFrom of attempts) {
    try {
      return requireFrom("playwright");
    } catch (error) {
      errors.push(error.message);
    }
  }
  throw new Error(`Playwright was not found. ${errors.join(" | ")}`);
}

function cleanText(value) {
  return String(value ?? "").replace(/\s+/g, " ").trim();
}

function sourceDomain(url) {
  try {
    return new URL(url).hostname.toLowerCase().replace(/^www\./, "");
  } catch {
    return "";
  }
}

function blockedRow(row, status, error) {
  return {
    rank: row.rank ?? "",
    mvp_scope: row.mvp_scope ?? "",
    brand: row.brand ?? "",
    model: row.model ?? "",
    registrations_ytd: row.registrations_ytd ?? "",
    source_url: row.source_url ?? "",
    final_url: "",
    http_status: "",
    source_domain: sourceDomain(row.source_url),
    render_status: status,
    render_error: error,
    source_hash: "",
    source_text_path: "",
    source_text_length: 0,
    source_text_preview: "",
    fetched_at: new Date().toISOString(),
    extraction_status: "blocked",
    notes: "No public data is populated from this row until rendered official source text is captured and validated.",
  };
}

function slug(value) {
  return String(value ?? "")
    .toLowerCase()
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 80);
}

async function renderRow(playwright, browser, row, timeoutMs) {
  const page = await browser.newPage({
    userAgent: "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 SwedishEVAdvisor/0.1",
    locale: "sv-SE",
  });
  try {
    const response = await page.goto(row.source_url, { waitUntil: "domcontentloaded", timeout: timeoutMs });
    await page.waitForLoadState("networkidle", { timeout: Math.min(timeoutMs, 12000) }).catch(() => {});
    const text = cleanText(await page.evaluate(() => document.body?.innerText ?? ""));
    const sourceHash = createHash("sha256").update(text, "utf8").digest("hex");
    const httpStatus = response?.status() ?? "";
    const pageNotFound =
      Number(httpStatus) >= 400 ||
      /sidan kunde inte hittas|page not found|404|not found/i.test(text.slice(0, 1200));
    const renderStatus = pageNotFound
      ? "rendered_page_not_found"
      : text.length >= MIN_TEXT_LENGTH
        ? "rendered_source_ready"
        : "rendered_text_too_short";
    let sourceTextPath = "";
    if (renderStatus === "rendered_source_ready") {
      await mkdir(DEFAULT_TEXT_DIR, { recursive: true });
      sourceTextPath = join(
        DEFAULT_TEXT_DIR,
        `${String(row.rank ?? "").padStart(2, "0")}-${slug(row.brand)}-${slug(row.model)}-${sourceHash.slice(0, 10)}.txt`,
      );
      await writeFile(sourceTextPath, `${text}\n`, "utf8");
    }
    return {
      rank: row.rank ?? "",
      mvp_scope: row.mvp_scope ?? "",
      brand: row.brand ?? "",
      model: row.model ?? "",
      registrations_ytd: row.registrations_ytd ?? "",
      source_url: row.source_url ?? "",
      final_url: page.url(),
      http_status: httpStatus,
      source_domain: sourceDomain(page.url()),
      render_status: renderStatus,
      render_error: renderStatus === "rendered_source_ready" ? "" : renderStatus.replace("rendered_", ""),
      source_hash: renderStatus === "rendered_source_ready" ? sourceHash : "",
      source_text_path: sourceTextPath,
      source_text_length: text.length,
      source_text_preview: text.slice(0, 500),
      fetched_at: new Date().toISOString(),
      extraction_status: renderStatus === "rendered_source_ready" ? "queued" : "blocked",
      notes: "Rendered official source text captured for AI extraction staging.",
    };
  } catch (error) {
    return blockedRow(row, "render_failed", error.message);
  } finally {
    await page.close().catch(() => {});
    void playwright;
  }
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const queueText = await readFile(args.queue, "utf8");
  let rows = parseCsv(queueText)
    .filter((row) => row.queue_status === "queued")
    .sort((a, b) => Number(a.priority || a.rank || 999) - Number(b.priority || b.rank || 999));
  if (Number.isFinite(args.limit)) rows = rows.slice(0, args.limit);

  const playwright = loadPlaywright();
  const launchAttempts = [
    { headless: true, channel: "msedge" },
    { headless: true, channel: "chrome" },
    { headless: true },
  ];
  let browser;
  const launchErrors = [];
  for (const options of launchAttempts) {
    try {
      browser = await playwright.chromium.launch(options);
      break;
    } catch (error) {
      launchErrors.push(`${options.channel ?? "bundled"}: ${error.message.split("\n")[0]}`);
    }
  }
  if (!browser) {
    throw new Error(`No browser runtime could be launched. ${launchErrors.join(" | ")}`);
  }
  const rendered = [];
  try {
    for (const row of rows) {
      rendered.push(await renderRow(playwright, browser, row, args.timeoutMs));
    }
  } finally {
    await browser.close().catch(() => {});
  }

  let existing = [];
  try {
    existing = parseCsv(await readFile(args.csvOut, "utf8"));
  } catch {}
  const merged = mergeRenderedRows(existing, rendered);

  await writeOutputs(merged, args.csvOut, args.jsonOut);
  const ready = rendered.filter((row) => row.render_status === "rendered_source_ready").length;
  const totalReady = merged.filter((row) => row.render_status === "rendered_source_ready").length;
  console.log(`Browser-render rows processed: ${rendered.length}`);
  console.log(`Rendered sources ready: ${ready}`);
  console.log(`Rendered sources ready total: ${totalReady}`);
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});

File: scripts/fetch_browser_rendered_sources.py
import argparse
import asyncio
import csv
import hashlib
import importlib.util
import json
import re
from datetime import datetime, timezone
from pathlib import Path
from urllib.parse import urlparse


ROOT = Path(__file__).resolve().parents[1]
DEFAULT_QUEUE = ROOT / "data/mvp/browser_render_queue.csv"
DEFAULT_CSV_OUT = ROOT / "data/extraction/browser_rendered_sources.csv"
DEFAULT_JSON_OUT = ROOT / "data/extraction/browser_rendered_sources.json"
MIN_TEXT_LENGTH = 500


def read_csv(path):
    if not path.exists():
        return []
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        return list(csv.DictReader(handle))


def write_csv(path, rows):
    path.parent.mkdir(parents=True, exist_ok=True)
    fieldnames = [
        "rank",
        "mvp_scope",
        "brand",
        "model",
        "registrations_ytd",
        "source_url",
        "final_url",
        "source_domain",
        "render_status",
        "render_error",
        "source_hash",
        "source_text_length",
        "source_text_preview",
        "fetched_at",
        "extraction_status",
        "notes",
    ]
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def write_json(path, rows):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(rows, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def clean_text(value):
    return re.sub(r"\s+", " ", value or "").strip()


def domain_from_url(url):
    return urlparse(url).netloc.lower().replace("www.", "")


def playwright_available():
    return importlib.util.find_spec("playwright") is not None


async def fetch_rendered_source(url, timeout_ms=30000):
    from playwright.async_api import async_playwright

    async with async_playwright() as runtime:
        browser = await runtime.chromium.launch(headless=True)
        page = await browser.new_page(
            user_agent="Mozilla/5.0 SwedishEVAdvisor/0.1",
            locale="sv-SE",
        )
        await page.goto(url, wait_until="networkidle", timeout=timeout_ms)
        text = clean_text(await page.locator("body").inner_text(timeout=timeout_ms))
        final_url = page.url
        await browser.close()
        return final_url, text


def blocked_row(row, status, error):
    return {
        "rank": row.get("rank", ""),
        "mvp_scope": row.get("mvp_scope", ""),
        "brand": row.get("brand", ""),
        "model": row.get("model", ""),
        "registrations_ytd": row.get("registrations_ytd", ""),
        "source_url": row.get("source_url", ""),
        "final_url": "",
        "source_domain": domain_from_url(row.get("source_url", "")),
        "render_status": status,
        "render_error": error,
        "source_hash": "",
        "source_text_length": 0,
        "source_text_preview": "",
        "fetched_at": datetime.now(timezone.utc).isoformat(),
        "extraction_status": "blocked",
        "notes": "No public data is populated from this row until rendered official source text is captured and validated.",
    }


async def render_row(row, timeout_ms):
    try:
        final_url, text = await fetch_rendered_source(row["source_url"], timeout_ms=timeout_ms)
        source_hash = hashlib.sha256(text.encode("utf-8")).hexdigest()
        text_length = len(text)
        status = "rendered_source_ready" if text_length >= MIN_TEXT_LENGTH else "rendered_text_too_short"
        return {
            "rank": row.get("rank", ""),
            "mvp_scope": row.get("mvp_scope", ""),
            "brand": row.get("brand", ""),
            "model": row.get("model", ""),
            "registrations_ytd": row.get("registrations_ytd", ""),
            "source_url": row.get("source_url", ""),
            "final_url": final_url,
            "source_domain": domain_from_url(final_url or row.get("source_url", "")),
            "render_status": status,
            "render_error": "" if status == "rendered_source_ready" else "source_text_too_short",
            "source_hash": source_hash if status == "rendered_source_ready" else "",
            "source_text_length": text_length,
            "source_text_preview": text[:500],
            "fetched_at": datetime.now(timezone.utc).isoformat(),
            "extraction_status": "queued" if status == "rendered_source_ready" else "blocked",
            "notes": "Rendered official source text captured for AI extraction staging.",
        }
    except Exception as error:
        return blocked_row(row, "render_failed", str(error))


async def build_rendered_sources(rows, limit=None, timeout_ms=30000, force_unavailable=False):
    queued = [row for row in rows if row.get("queue_status") == "queued"]
    queued.sort(key=lambda row: int(row.get("priority") or row.get("rank") or 999))
    if limit is not None:
        queued = queued[:limit]

    if force_unavailable or not playwright_available():
        return [
            blocked_row(
                row,
                "blocked_missing_playwright",
                "Playwright is not installed in this workspace; install/enable a browser runtime before rendering.",
            )
            for row in queued
        ]

    return [await render_row(row, timeout_ms) for row in queued]


def main():
    parser = argparse.ArgumentParser(description="Fetch official manufacturer pages that require browser rendering.")
    parser.add_argument("--queue", default=str(DEFAULT_QUEUE))
    parser.add_argument("--csv-out", default=str(DEFAULT_CSV_OUT))
    parser.add_argument("--json-out", default=str(DEFAULT_JSON_OUT))
    parser.add_argument("--limit", type=int, default=None)
    parser.add_argument("--timeout-ms", type=int, default=30000)
    args = parser.parse_args()

    rows = read_csv(Path(args.queue))
    rendered = asyncio.run(build_rendered_sources(rows, limit=args.limit, timeout_ms=args.timeout_ms))
    write_csv(Path(args.csv_out), rendered)
    write_json(Path(args.json_out), rendered)

    ready = sum(1 for row in rendered if row["render_status"] == "rendered_source_ready")
    print(f"Browser-render rows processed: {len(rendered)}")
    print(f"Rendered sources ready: {ready}")
    if ready == 0:
        print("No rendered source text was captured. This is expected if Playwright/browser runtime is unavailable.")


if __name__ == "__main__":
    main()
